### PR TITLE
[SPARK-25553][BUILD] Add EmptyInterpolatedStringChecker to scalastyle config

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

 [EmptyInterpolatedStringChecker](http://www.scalastyle.org/rules-dev.html#org_scalastyle_scalariform_EmptyInterpolatedStringChecker) used for check for empty string interpolations.
This feature is very useful for us. This PR add it to `scalastyle-config.xml` and fix all empty interpolated string issue.

## How was this patch tested?

manual tests

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
The usual question here is, is it worth the extra spurious merge conflicts that minor improvements like this will cause? (Also, does the scala compiler actually do something different for `s"foo"` vs `"foo"`? I tried with `javap` and it looks like it treats both as just a string.)

I am mostly OK with cleaning this up, but would have to be for the time when `master` is bound for Spark 3.0. I think we should clean up legacy `.format(...)` calls too. I am not as worried about cases here where there is no actual runtime effect.

**Comment by HyukjinKwon** [comment]
Exactly same opinion with Sean's.

**Comment by dongjoon-hyun** [comment]
We had better update the code when we touch it.

**Comment by wangyum** [comment]
What do we do next?
1. Change `level` to `warning`and commit all changes. In this case `dev/scalastyle` can pass, but IDEA will have a warning:
 image
2. Remove `EmptyInterpolatedStringChecker` and commit all changes.
3. Just close this PR.

**Comment by dongjoon-hyun** [comment]
For me, this PR is too intrusive to merge at this stage.

**Comment by wangyum** [comment]
@srowen @maropu Do you think we can add it for Spark 3.0?

**Comment by HyukjinKwon** [comment]
I'm fine. WDYT guys?

**Comment by dongjoon-hyun** [comment]
It's still intrusive to me. But, 3.0.0 might be the last candidate to do this. I thought this is a reasonable request. I'll not be against this.

cc @cloud-fan , @gatorsmile , @rxin .

**Comment by cloud-fan** [comment]
SGTM

**Comment by maropu** [comment]
I'm fine, too.

File: core/src/main/scala/org/apache/spark/deploy/rest/StandaloneRestServer.scala [hunk 1]
```diff
@@ -91,7 +91,7 @@ private[rest] class StandaloneStatusRequestServlet(masterEndpoint: RpcEndpointRe
   protected def handleStatus(submissionId: String): SubmissionStatusResponse = {
     val response = masterEndpoint.askSync[DeployMessages.DriverStatusResponse](
       DeployMessages.RequestDriverStatus(submissionId))
-    val message = response.exception.map { s"Exception from the cluster:\n" + formatException(_) }
+    val message = response.exception.map { "Exception from the cluster:\n" + formatException(_) }
```
[the remark this answers is not in the document]
Comment (wangyum) [reply core/src/main/scala/org/apache/spark/deploy/rest/StandaloneRestServer.scala]: Do you mean?
```scala
val message = response.exception.map { s"Exception from the cluster:\n" + formatException(_) }
```
->
```scala
val message = response.exception.map { s"Exception from the cluster:\n ${formatException(_)}" }
```

**Comment by maropu** [comment]
retest this please

**Comment by srowen** [comment]
Whew, that's a lot of files! I'm hesitant to do this. I checked if the standard interpolator special-cases a string with no replacements, but doesn't look like it does. While the overhead is small, it may be worth avoiding this all over the code. I'm OK with it for 3.0

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by felixcheung** [comment]
can you please check appveyor build clean on this change? https://ci.appveyor.com/project/ApacheSoftwareFoundation/spark/builds/26773731 the last build failed, the next commit you push should trigger it again

**Comment by rxin** [comment]
It seemed like the cost of this is very high (merge conflict everywhere)
and syntactically it doesn’t bring much benefit. Why do this one?

**Comment by wangyum** [comment]
How about set `level="warn"`:
```xml
<check level="warn" class="org.scalastyle.scalariform.EmptyInterpolatedStringChecker" enabled="true"></check>
```
The scalastyle check can pass and we can see the warning in IDEA:
image

**Comment by HyukjinKwon** [comment]
If we turn to `warn`, then we won't have to change a lot of instances here, Yes.
But one concern is though then it will show 1000 lines of warnings about the style. One way to workaround this 1000 lines warnings, we could disable this rule when we actually build (see for instance https://github.com/apache/spark/commit/761fc46c70f15ffc3f35ff87b6a5c9192af1c55f).

The Intellij actually looks useful to avoid excessive (and useless) string interpolation, which people still nitpick in PRs time to time.

So, in this way, we could be able to the style only in IntelliJ and does not affect anything like build or existing codes.

But not sure if guys here like this idea given the complexity, and .. 3.0.0 might be the last candidate to do this .. maybe at least with few years.

**Comment by srowen** [comment]
I would not enforce it only in IJ... do it or don't, I say. I think if we bite the bullet and make this change, it needs to be enforced with style checker errors. I'm OK with it for 3.0.

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
All tests passed.

**Comment by dongjoon-hyun** [comment]
Hi, All. 

Give the current status, I believe @wangyum can resolve the conflicts and merge this as a committer. Is there something for us to consider more?

**Comment by HyukjinKwon** [comment]
There seems, I suspect, one objection at https://github.com/apache/spark/pull/22570#issuecomment-522299885 . If this one gets deadlock, closing is an option to I believe.

**Comment by srowen** [comment]
I'm OK dropping this on the grounds that it makes backports harder, even though this is for Spark 3. The argument for the change is a tiny bit of performance, but it's probably negligible except perhaps in a few hot paths. I wonder if there is a small subset of these changes that look like they could matter at all for perf and just fix those?

**Comment by dongjoon-hyun** [comment]
Yep. I agree that closing is one of the option, @HyukjinKwon and @srowen . And, another option in my mind is applying this rule in `master` and `branch-2.4` together. It will reduce the change of conflicts, @rxin .

As of today, we only have a historically minimum number of live branch `master` and `branch-2.4`. This is a good timing to do this. Also, the logical sync is not hard to do because it's a maven rule.

Anyway, shall we make a decision since this is open too long due to non-technical reason? We don't merge if we have one -1.

cc @rxin , @gatorsmile , @cloud-fan , @maropu , @srowen , @HyukjinKwon , @wangyum .

**Comment by dongjoon-hyun** [comment]
By default, if there is no comment updates in this PR in next a few days, please close this, @wangyum .

**Comment by HyukjinKwon** [comment]
> another option in my mind is applying this rule in master and branch-2.4 together. It will reduce the change of conflicts.

This sounds logical to me.

**Pull request closed without merging**